platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/heartbeats/HeartbeatProtocol.java

```diff
@@ -0,0 +1,147 @@
+import com.swirlds.platform.network.NetworkMetrics;
+import com.swirlds.platform.network.NetworkProtocolException;
+import com.swirlds.platform.network.protocol.Protocol;
+import com.swirlds.platform.network.unidirectional.UnidirectionalProtocols;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * Sends a heartbeat to the other node and measures the time it takes to receive a response.
+ */
+public class HeartbeatProtocol implements Protocol {
+    /**
+     * ID of the peer
+     */
+    private final NodeId otherId;
+
+    /**
+     * The last time a heartbeat protocol was executed
+     */
+    private Instant lastHeartbeatTime = Instant.MIN;
+
+    /**
+     * The period at which the heartbeat protocol should be executed
+     */
+    private final Duration heartbeatPeriod;
+
+    /**
+     * Network metrics, for recording roundtrip heartbeat time
+     */
+    private final NetworkMetrics networkMetrics;
+
+    /**
+     * Constructor
+     *
+     * @param otherId         ID of the peer
+     * @param heartbeatPeriod The period (milliseconds) at which this protocol should execute
+     * @param networkMetrics  Network metrics, for recording roundtrip heartbeat time
+     */
+    public HeartbeatProtocol(
+            final @NonNull NodeId otherId,
+            final @NonNull Duration heartbeatPeriod,
+            final @NonNull NetworkMetrics networkMetrics) {
+
+        this.otherId = ArgumentUtils.throwArgNull(otherId, "otherId");
+        this.heartbeatPeriod = ArgumentUtils.throwArgNull(heartbeatPeriod, "heartbeatPeriod");
+        this.networkMetrics = ArgumentUtils.throwArgNull(networkMetrics, "networkMetrics");
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Returns true if the last heartbeat protocol was started more than {@link #heartbeatPeriod} ago
+     */
+    @Override
+    public boolean shouldInitiate() {
+        return Duration.between(lastHeartbeatTime, Instant.now()).compareTo(heartbeatPeriod) >= 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldAccept() {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean acceptOnSimultaneousInitiate() {
+        return true;
+    }
+
+    /**
+     * Reads a byte from the socket
+     *
+     * @param connection the connection to read from
+     * @param startTime  the start time measurement for determining roundtrip time
+     * @throws IOException              if there is an error reading from the socket
+     * @throws NetworkProtocolException if the byte read is not a heartbeat or heartbeat ack
+     */
+    private void readByte(final @NonNull Connection connection, final long startTime)
+            throws IOException, NetworkProtocolException {
+
+        final byte readByte = connection.getDis().readByte();
+
+        if (readByte == ByteConstants.HEARTBEAT) {
```

Comment:
should this be checking for `UnidirectionalProtocols.HEARTBEAT.getInitialByte()`?